lxd/storage/drivers/driver_dir.go

```diff
@@ -0,0 +1,864 @@
+// filler function.
+func (d *dir) CreateVolume(vol Volume, filler func(path string) error, op *operations.Operation) error {
+	if vol.ContentType != ContentTypeFS {
+		return fmt.Errorf("Content type not supported")
+	}
+
+	volPath := vol.MountPoint()
+
+	// Get the volume ID for the new volume, which is used to set project quota.
+	volID, err := d.getVolID(vol.Type, vol.Name)
+	if err != nil {
+		return err
+	}
+
+	err = os.MkdirAll(volPath, 0711)
+	if err != nil {
+		return err
+	}
+
+	revertPath := true
+	defer func() {
+		if revertPath {
+			d.deleteQuota(volPath, volID)
+			os.RemoveAll(volPath)
+		}
+	}()
+
+	// Initialise the volume's quota using the volume ID.
+	err = d.initQuota(volPath, volID)
+	if err != nil {
+		return err
+	}
+
+	// Set the quota if specified in volConfig or pool config.
+	err = d.setQuota(volPath, volID, vol.Config["size"])
+	if err != nil {
+		return err
+	}
+
+	if filler != nil {
+		err = filler(volPath)
+		if err != nil {
+			return err
+		}
+	}
+
+	revertPath = false
+	return nil
+}
+
+// MigrateCustomVolume sends a volume for migration.
+func (d *dir) MigrateCustomVolume(vol Volume, conn *websocket.Conn, volSrcArgs migration.VolumeSourceArgs, op *operations.Operation) error {
+	if volSrcArgs.MigrationType.FSType != migration.MigrationFSType_RSYNC {
+		return fmt.Errorf("Migration type not supported")
+	}
+
+	bwlimit := d.config["rsync.bwlimit"]
+
+	for _, snapName := range volSrcArgs.Snapshots {
+		fullSnapName := fmt.Sprintf("%s%s%s", vol.Name, shared.SnapshotDelimiter, snapName)
+		snapshot := NewVolume(d, d.name, vol.Type, vol.ContentType, fullSnapName, nil)
+		wrapper := migration.ProgressReader(op, "fs_progress", fullSnapName)
+		ourMount, err := snapshot.Mount(op)
+		if err != nil {
+			return err
+		}
+
+		path := shared.AddSlash(snapshot.MountPoint())
+		err = rsync.Send(fullSnapName, path, conn, wrapper, volSrcArgs.MigrationType.Features, bwlimit, d.state.OS.ExecPath)
+		if err != nil {
+			if ourMount {
+				snapshot.Unmount(op)
+			}
+
+			return err
+		}
+
+		if ourMount {
+			snapshot.Unmount(op)
+		}
+	}
+
+	wrapper := migration.ProgressReader(op, "fs_progress", vol.Name)
+	ourMount, err := vol.Mount(op)
+	if err != nil {
+		return err
+	}
+
+	path := shared.AddSlash(vol.MountPoint())
+	err = rsync.Send(vol.Name, path, conn, wrapper, volSrcArgs.MigrationType.Features, bwlimit, d.state.OS.ExecPath)
+	if err != nil {
+		if ourMount {
+			vol.Unmount(op)
+		}
+
+		return err
+	}
+
+	if ourMount {
+		vol.Unmount(op)
+	}
+
+	return nil
+}
+
+// CreateVolumeFromMigration creates a volume being sent via a migration.
+func (d *dir) CreateVolumeFromMigration(vol Volume, conn *websocket.Conn, volTargetArgs migration.VolumeTargetArgs, op *operations.Operation) error {
+	if vol.ContentType != ContentTypeFS {
+		return fmt.Errorf("Content type not supported")
+	}
+
+	if volTargetArgs.MigrationType.FSType != migration.MigrationFSType_RSYNC {
```

Comment:
We're going to have to support block eventually though, we'll effectively do that by supporting MigrationFSType_DD which will get us a simple stream of bytes from the source and then will have to pipe that through qemu-img or something to write it as qcow2 to a file on disk.

We'll need that to be able to move VMs around.